### ISSUE TITLE
get compensation voltage from mzXML files

### ISSUE THREAD

**Issue by jgmeyerucsd** [sneumann/mzR#195]
I'm was planning to try and contribute a method to get the value of "compensationVoltage" from the scan header (relevant for the new thermo bolt-on FAIMS device), but I'm realizing after reading your code that I don't understand the formal structure of R packages and methods. I read the latest commit where you added the SWATH-related values: https://github.com/sneumann/mzR/issues/193

Could you please also add a method that grabs this value? Also please let me know if what I think looks like calls to Rcpp syntax work for grabbing arbitrary values from the scan header? I suppose I will see the answer to the latter when I compare this update to the referenced issue. 

Thank you

**Comment by jorainer** [comment]
So far we unfortunately don't have any code to grab arbitrary values from the scan header. Most of the code we have relies on the proteowizard C++ library to extract and return the specific information from the mzML files. Could you eventually provide the official CV term for the *compensationVoltage* (MS:...)?

**Comment by jgmeyerucsd** [comment]
Ok I'll work with the prkteowizard people first to figure that part out.
I'm not sure if there is an official controlled vocab for this yet as it
was only released commercially a month or two ago. Who do I talk to about
adding that?

**Comment by jorainer** [comment]
Oh, I see. That I don't know. And even if it was added as a CV term to proteowizard - it's quite a struggle to include proteowizard into the `mzR` package, so the version which is included at present is not the newest one... 

Alternatively you might try to implement a function based on the `xml2` package that extracts that field?

**Comment by lgatto** [comment]
Yes, it's possible with `xml2` - I have done this in the past, and there still might be code in `MSnbase` that parses the `mzML` files to extract specific CV terms using `xml2`. I'll try to find it later, when I've got time.

**Comment by sneumann** [comment]
Hi,
do you have a small example mzML file somewhere ?
The term needs to be in https://github.com/HUPO-PSI/psi-ms-CV/blob/master/psi-ms.obo
That file is best browsed via https://www.ebi.ac.uk/ols/ontologies/ms
Once the term is there, we need to find if pwiz can get this out of the mzML. 
I have a pwiz update for mzR in the works,
https://github.com/sneumann/mzR/tree/feature/updatePwiz20190301
which currently fails to build/run to satisfaction.
So timing would be good to try to get that in. 
Yours, Steffen

**Comment by jgmeyerucsd** [comment]
Regarding the presence of the term in psi-ms-CV, I did not find it there when searching the ebi.ac.uk link you provided. I opened an issue on their github here:
https://github.com/HUPO-PSI/psi-ms-CV/issues/22

However, I did convert my raw files with proteowizard and the term is present in the mzXML file, so proteowizard accommodates including the term somehow. I'm not sure if it dumps everything from the scan header or how proteowizard works. 

As for the small example file, you can find a pair of .raw and proteowizard-converted .mzxml files here:
https://uwprod-my.sharepoint.com/:f:/g/personal/jgmeyer2_wisc_edu/EnT4A4jHk-1LoC6OZ8UMwosBIGW0z5f-Z_9wJPcsw5QgtA?e=X4bAxf

from the data used in this publication:
DOI: 10.1021/acs.analchem.8b02233

Meanwhile I found a python package called pyteomics allows me to grab the whole scan header from /mzXML into a dictionary, and I've translated most of what I was building into python this morning, so I don't really need this anymore but others might in the future. That said, it might be useful if mzR had similar functionality to enable grabbing the whole scan header, which would prevent additional requests for "please add method to get _foo_ from the scan header".

**Comment by jgmeyerucsd** [comment]
@edeutsch said the term is already there:
https://www.ebi.ac.uk/ols/ontologies/ms/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FMS_1001581

**Comment by sneumann** [comment]
Hi, the sharepoint link does not work for me. Would you have mzML as well ? Yours, Steffen

**Comment by jgmeyerucsd** [comment]
Try this one please:

https://drive.google.com/drive/folders/13tjxUtuhtKaXoTr6FZKWhmX_PBBQtww_?usp=sharing

**Comment by sneumann** [comment]
Great, here it is:
```
        <spectrum index="0" id="controllerType=0 controllerNumber=1 scan=1" defaultArrayLength="804" dataProcessingRef="pwiz_Reader_conversion">
          <cvParam cvRef="MS" accession="MS:1000579" name="MS1 spectrum" value=""/>
          <cvParam cvRef="MS" accession="MS:1000511" name="ms level" value="1"/>
          <cvParam cvRef="MS" accession="MS:1000130" name="positive scan" value=""/>
          <cvParam cvRef="MS" accession="MS:1000127" name="centroid spectrum" value=""/>
          <cvParam cvRef="MS" accession="MS:1000504" name="base peak m/z" value="397.918121337891" unitCvRef="MS" unitAccession="MS:1000040" unitName="m/z"/>
          <cvParam cvRef="MS" accession="MS:1000505" name="base peak intensity" value="31539.21875" unitCvRef="MS" unitAccession="MS:1000131" unitName="number 
of detector counts"/>
          <cvParam cvRef="MS" accession="MS:1000285" name="total ion current" value="2.43293e05"/>
=>        <cvParam cvRef="MS" accession="MS:1001581" name="FAIMS compensation voltage" value="-10.0"/>
          <cvParam cvRef="MS" accession="MS:1000528" name="lowest observed m/z" value="297.016571249051" unitCvRef="MS" unitAccession="MS:1000040" unitName="m/z"/>
          <cvParam cvRef="MS" accession="MS:1000527" name="highest observed m/z" value="1363.585427792357" unitCvRef="MS" unitAccession="MS:1000040" unitName="m/z"/>
          <cvParam cvRef="MS" accession="MS:1000796" name="spectrum title" value="20feb2018_60cv_60min_CV_10.1.1. File:&quot;20feb2018_60cv_60min_CV_10.raw&quot;, NativeID:&quot;controllerType=0 controllerNumber=1 scan=1&quot;"/>
```
and the value is -10 for all scans:
```
sneumann@msbi:~$ grep "MS:1001581"  ~/Downloads/20feb2018_60cv_60min_CV_10_centroid.mzML  | uniq -c
  14371           <cvParam cvRef="MS" accession="MS:1001581" name="FAIMS compensation voltage" value="-10.0"/>
```

**Comment by jgmeyerucsd** [comment]
Thanks for looking at this. I know the value for this file is static, but in some analysis we use multiple CV values and I wanted to grab it easily in R as part of my analysis.
